heartbeat/monitors/wrappers/monitorstate/esloader.go

```diff
@@ -0,0 +1,115 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package monitorstate
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+
+	"github.com/elastic/elastic-agent-libs/logp"
+	"github.com/elastic/elastic-agent-libs/mapstr"
+	"github.com/elastic/go-elasticsearch/v8"
+	"github.com/elastic/go-elasticsearch/v8/esapi"
+
+	"github.com/elastic/beats/v7/heartbeat/config"
+	"github.com/elastic/beats/v7/heartbeat/esutil"
+	"github.com/elastic/beats/v7/heartbeat/monitors/stdfields"
+)
+
+func MakeESLoader(esc *elasticsearch.Client, indexPattern string, beatLocation *config.LocationWithID) StateLoader {
+	if indexPattern == "" {
+		// Should never happen, but if we ever make a coding error...
+		logp.L().Warn("ES state loader initialized with no index pattern, will not load states from ES")
+		return NilStateLoader
+	}
+	return func(sf stdfields.StdMonitorFields) (*State, error) {
+		queryMustClauses := []mapstr.M{
+			{
+				"match": mapstr.M{"monitor.id": sf.ID},
+			},
+			{
+				"match": mapstr.M{"monitor.type": sf.Type},
+			},
+			{
+				"exists": mapstr.M{"field": "state"},
+			},
+			{
+				// Only search the past 6h of data for perf, otherwise we reset the state
+				// Monitors should run more frequently than that.
+				"range": mapstr.M{"@timestamp": mapstr.M{"gt": "now-6h"}},
```

Comment:
Shouldn't we look at record counts instead of doing by range? Reason being this varies depending on how frequent individual monitors are scheduled to run.

Reply:
data streams / time series indexes really should always be paired with a time range if possible since it dramatically improves efficiency. ES can discard rotated indexes entirely if they don't contain that range. So, rather than having to examine say 365 indices with daily rotation it'd only query one.

Since 6h is well beyond the interval the UI supports, or a reasonable user would choose, I thought this made sense. Glad to up it to 12 or 24h but 6 seems quite large IMHO.

Reply:
I think by record counts I think you meant by timestamp descending, which this query does do.

Reply:
Thanks for the details Andrew. It makes sense, I was going for like
- Don't care about the range
- Just get the entries that we are interested in. 

But I did not know about the efficiency improvement we get with the time ranges in ES. ++ 6 hours is way too large.

Reply:
Yeah, unfortunately ES has no way, without a timestamp, to just get the latest AFAIK. My understanding is that if you want to get the latest it gets the most recent document from every index, then finds the most recent one. With a time range it excludes all but the most recent index